cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaOffsetFinder.java

```diff
@@ -0,0 +1,284 @@
+
+  public KafkaOffsetFinder(BrokerService brokerService, KafkaConsumer consumer, String topic) throws IOException {
+    this.brokerService = brokerService;
+    this.consumer = consumer;
+    this.topic = topic;
+    this.serializer = new LoggingEventSerializer();
+  }
+
+  public long getMatchingOffset(final Checkpoint checkpoint, final int partition) {
+    if (checkpoint.getNextOffset() < 0) {
+      return checkpoint.getNextOffset();
+    }
+
+    // First try to check whether the message with the offset in the given checkpoint has the same timestamp.
+    // If the message has the same timestamp as in the checkpoint, directly return the checkpoint's offset.
+    final AtomicInteger matchFound = new AtomicInteger(0);
+    Cancellable initCancel = null;
+    try {
+      initCancel = consumer.prepare().add(topic, partition, checkpoint.getNextOffset())
+        .consume(new KafkaConsumer.MessageCallback() {
+          @Override
+          public long onReceived(Iterator<FetchedMessage> messages) {
+            if (!messages.hasNext()) {
+              return EARLIEST;
+            }
+            FetchedMessage message = messages.next();
+            long offset = message.getOffset();
+            ILoggingEvent event;
+            try {
+              event = serializer.fromBytes(message.getPayload());
+            } catch (IOException e) {
+              LOG.warn("Message with offset {} in topic {} partition {} is ignored because of failure to decode.",
+                       offset, topic, partition, e);
+              return message.getNextOffset();
+            }
+            long timestamp = event.getTimeStamp();
+            if (offset == checkpoint.getNextOffset()) {
+              matchFound.set(timestamp == checkpoint.getMaxEventTime() ? TIME_MATCH : TIME_NOT_MATCH);
+            }
+            return message.getNextOffset();
+          }
+
+          @Override
+          public void finished() {
+            // no-op
+          }
+        });
+      Tasks.waitFor(true, new Callable<Boolean>() {
+        @Override
+        public Boolean call() throws Exception {
+          return matchFound.get() != 0;
+        }
+      }, 1500, TimeUnit.MILLISECONDS);
+
+      if (matchFound.get() == TIME_MATCH) {
+        initCancel.cancel();
+        return checkpoint.getNextOffset();
+      }
+
+    } catch (Exception e) {
+      LOG.trace("Given checkpoint {} has mismatching offset and timestamp. Continue to search for the correct offset" +
+                  "with the given timestamp.", checkpoint, e);
+    }
+    if (initCancel != null) {
+      initCancel.cancel();
+    }
+
+    // Get BrokerInfo for constructing kafka.javaapi.consumer.SimpleConsumer in OffsetFinderCallback
+    BrokerInfo brokerInfo = brokerService.getLeader(topic, partition);
+    if (brokerInfo == null) {
+      return EARLIEST; // Return earliest offset if brokerInfo is null
+    }
+    final BlockingQueue<Long> offsetQueue = new LinkedBlockingQueue<>();
```

Comment:
This will depend on how the new LogSaver will be implemented. LogSaver currently iterates through all the `KafkaLogProcessor` in each partition to find out the lowest saved checkpoint first, and then start reading from the lowest checkpoint. Therefore, I use `KafkaLogProcessor#resetCheckpoint` to reset the check points first before getting the lowest saved checkpoint. A KafkaMessageCallback wrapper cannot control where LogSaver starts to read.